Add half-cones and Ipelet

freud14 opened the pull request:
Hi, I made a small Ipelet using the Cone-Based Spanners package. I also included a way to construct half-theta and half-yao graphs. Thank you.

Small-feature page [here](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Cone_spanner_half_cones).

Note: pre-approved 20 dec. 2016

### TODO
- [x] update `changes.html`

ws4u commented:
Thank @freud14 for the contribution. I looked at the code for half yao/theta graph construction. My comments are as follows:

- The API is properly modified by adding a default parameter to indicate full or half construction. I think this API modification is necessary, and kept at minimum.
- The code for the actual construction is concise. I didn't download and run the code, but according to my reading of the code, I think the code can correctly construct half yao/theta graphs. 
- If the code is to be merged into master, the user manual for the cone-based spanners should be updated to document the updated API.
- There is a small typo in the code comments: wheter -> whether

For the added Ipelet, I'm not good at it. However, @sloriot may have time to comment on it.

freud14 commented:
Thank you @ws4u for your response. I made a commit to fix the small typo. For the user manual, what do you suggest? IMO, a small mention in the Definitions Section would do. Thank you.

freud14 commented:
Hi,

I notice that when two points are on the horizontal line, because of the imprecision of the Exact_predicates_inexact_constructions_kernel, I do not get what I am supposed to. I tested with the Exact_predicates_exact_constructions_kernel_with_root_of and it seems to work fine. I read [there](http://doc.cgal.org/latest/CGAL_ipelets/index.html#Chapter_CGAL_Ipelets) that it is fine to use this kernel in an Ipelet, but I get a linking error in Ipe when I try to use the Ipelet unless I do `export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib/` before opening Ipe. Any idea?

Furthermore, I just made some tests and it seems that the documentation is inaccurate. It says that a point on l_i is part of the cone c(u, i). However, according to my tests in Ipe, it's the points on l_{i+1} that seem to belong to c(u, i).

Thank you.

Frédérik

ws4u commented:
@freud14 For your second question, I'll experiment myself and examine the code again. This week I'm a little busy, but I'll get back to you next week. Thanks for your feedback.

lrineau commented:
@sloriot As the unofficial maintainer of our ipelets, can you please check this PR?

lrineau commented:
Its seems the API and documentation of `Cone_spanners_2` is amended. In my opinion there should be a corresponding small feature page.

ws4u commented:
@freud14 I tested and examined my code again. You are correct. In the current implementation, l_{i+1}, not l_i, belongs to c(u, i). To get an implementation that l_i belongs to c(u,i), the following two lines of code in "Less_by_direction_2.h" 
`  // define a rotation of counter clockwise 90`
`  Transformation ccw90(0, -1, 1,  0);`
should be changed to:
`  // define a rotation of clockwise 90`
`  Transformation cw90(0, 1, -1,  0);`
This will perfectly fix the problem. I've tested about it. If you have further questions, please feel free to let me know.

freud14 commented:
@ws4u Thank you for your answer. The question is: do we want to change the implementation or the documentation?

ws4u commented:
@freud14 I will create a pull request to change the implementation, and also add code comments to the implementation on what to do if users want l_{i+1} on c(u,i).

sloriot commented:
The ipelet compiles and run. The code looks good too.

freud14 commented:
@sloriot I have a small problem that I mentionned in this comment: https://github.com/CGAL/cgal/pull/1673#issuecomment-260390903 Do you have an answer for it? Thank you.

sloriot commented:
@freud14  I guess you have a link issue since Core is needed when `Exact_predicates_exact_constructions_kernel_with_root_of` is used.

freud14 commented:
@sloriot Hmm that's weird. It wasn't working correctly, but now it works. Thank you. I just committed the change of kernel if you want to verify it.

sloriot commented:
Runs correctly here too.

freud14 commented:
> In my opinion there should be a corresponding small feature page.

@lrineau Do you want me to do that? Do you have instructions for that?

sloriot commented:
I suggest that you do the modification asked by @ws4u in the user manual (updating the definition section and mention half-XXX in the description in `PackagesDescription.txt`). When @ws4u is happy with the modifications, we'll deal with the small feature as you don't have access to our internal resources (this is a human readable page summarizing API change and doc changes).
I have a naive question (that also applies to Yao-graph) after reading the Wikipedia page of Theta-graph. I guess you compute one of _even_ and _odd_ Theta-graph (_even_ IIUC). Why not offering the possibility to also compute the complement?

freud14 commented:
> Why not offering the possibility to also compute the complement?

That's a good idea. So, instead of a boolean, it could be an enum with items {EVEN_CONES, ODD_CONES, ALL_CONES} ?

sloriot commented:
Looks good to me. @ws4u what do you think?

ws4u commented:
@sloriot Looks good to me as well. For the Ipelet part, where should we document it, in both CGAL Ipelets chapter and Cone-based Spanners chapter?

@freud14 In modifying `PackagesDescription.txt`, feel free to ask me if you have any questions. Please add you as an author in the user manual.

sloriot commented:
The ipelet should be mentioned in the user manual of the ipelets.

freud14 commented:
So, I added the choice for even or odd cones and I added all the necessary documentation. However, I don't know why but the elements of the enum don't seem to display in the doc. I don't know what I did wrong. I tried to copy the enum in the CGAL/enum.h files of Kernel_23 and it was working fine so I'm wondering if there is some configuration to add to the Cone_spanners_2 package.

ws4u commented:
Thanks @freud14. I suggest the following changes to your submission:

- Rename the file "Cone_spanners_2_enum.h" to "Cone_spanners_enum_2.h"
- Rename the enum type `Half` to `Cones_selected`
- Rename the member variable `half` to `cones_selected`
- Change the comment `/* Store whether this is an half-theta graph */` to `/* Store whether even, odd or all cones are selected to construct graph */`
- The user manual is converted from the "Cone_spanners_2.txt" file. I'm sorry for mentioning it wrongly as "PackageDescription.txt"  in my last comment. Pls add a new subsection into this file after the subsection titled "Constructing Graphs Exactly or Inexactly and Generating Gnuplot Files". This new subsection should be titled sth like "Constructing Half Theta or Yao Graphs". In this new subsection, pls briefly describe the  following:
-- By default, a full theta or yao graph is constructed.
-- To construct a half theta or yao graph, users should include "Cone_spanners_enum_2.h" in the cpp file, and supply one more parameter when initializing the functor.
-- Adding a code snippet for this will be OK. No need to include an entire file.
- Add authorship to the files where you make contributions.

For the above suggestions, you don't need to follow them strictly, but can have your own creative ideas. For the above renamings, pls wait for @sloriot 's comment before carrying them out.

ws4u commented:
@freud14, @sloriot  To make the elements of the enum display in the doc, a compilation on the doc is needed. I'm not sure whether this is available to public contributors.

freud14 commented:
@ws4u Just so you know, I already added a small paragraph in the Definitions Section. I will add a section like you suggest.

freud14 commented:
I'm already able to compile the doc. When I click on the type of my parameter, it just stays on this [page](http://imgur.com/zH3t6mL.png) and doesn't display the elements of the enum.

ws4u commented:
Can you try adding comments for each element?

sloriot commented on CGAL_ipelets/demo/CGAL_ipelets/cone_spanners.cpp:
```diff
@@ -0,0 +1,148 @@
+public:
+  Cone_spanners_ipelet()
+    :CGAL::Ipelet_base<Kernel,6>("Cone Spanners",labels,hmsg){}
+  void protected_run(int);
```
6 -> 7 I guess

sloriot commented on Cone_spanners_2/include/CGAL/Cone_spanners_2_enum.h:
```diff
@@ -0,0 +1,38 @@
+
+   \brief An enum of the types of cone spanners.
+   */
+  enum Half { EVEN_CONES = 0, ODD_CONES = 1, ALL_CONES = 2 };
```
The name of the enum is too generic.

sloriot commented on Cone_spanners_2/include/CGAL/Construct_theta_graph_2.h:
```diff
@@ -90,10 +94,13 @@ class Construct_theta_graph_2 {
                    cones. This allows arbitary rotations of the rays that divide
                    the plane.  (default: positive x-axis)
+     \param half_theta  Indicates whether all the cones are used or just half of
+                        them (default: all cones).
```
no need to mention the default

sloriot commented on Cone_spanners_2/include/CGAL/Construct_yao_graph_2.h:
```diff
@@ -85,10 +89,13 @@ class Construct_yao_graph_2 {
                    cones. This allows arbitary rotations of the rays that divide
                    the plane.  (default: positive x-axis)
+     \param half_yao  Indicates whether all the cones are used or just half of
+                      them (default: all cones).
```
No need to mention the default

sloriot commented on Cone_spanners_2/include/CGAL/Cone_spanners_2_enum.h:
```diff
@@ -0,0 +1,38 @@
+//
+// Authors: Weisheng Si, Quincy Tse
+
+/*! \file Cone_spanners_2_enum.h
```
Rename the file `Cone_spanners_enum_2.h`

freud14 commented:
I did the modifications asked. I added comments on each element of the enum so that they can display in the doc. However, when trying to reference the enum with quotes (\`), the link is not made in the doc. I don't know why.

freud14 commented:
Done.

ws4u commented:
Great, I'm ok with the doc now. For the back quote (`), it doesn't work occasionally for me as well. I don't know why.

lrineau commented:
@sloriot might know how to fix the auto-linking in the documentation.

freud14 commented:
Hi, is there anything remaining I have to do? Thank you.

sloriot commented:
Sorry I forgot to submit the small feature. I'll do it right away.

sloriot commented on Cone_spanners_2/include/CGAL/Construct_yao_graph_2.h:
```diff
@@ -16,7 +16,7 @@
 //
 //
-// Authors: Weisheng Si, Quincy Tse
+// Authors: Weisheng Si, Quincy Tse, Frédérk Paradis
```
typo here

freud14 replied:
Fixed.

sloriot commented:
TODO
- [ ] update `changes.html`

sloriot commented:
@ws4u you can check out and pre-approved the [small feature page](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Cone_spanner_half_cones).

ws4u commented:
@sloriot I have no problem with the small feature page.

lrineau commented:
There are `-Wreorder` warnings with g++:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.10-Ic-81.shtml#Cone_spanners_2
```
.../include/CGAL/Construct_theta_graph_2.h:103:48: warning: field 'rays' will be initialized after field 'cones_choice' [-Wreorder]
                            ): cone_number(k), rays(std::vector<Direction_2>(k)), cones_choice(cones_selected)
                                               ^
```

freud14 commented:
Fixed. Btw, do you receive a notification when I push a commit?

lrineau commented:
Thanks. Yes, I watch the whole CGAL/cgal repo, and that means all pull-requests. You can also see me in the list of participants of this discussion.

This PR will be re-tested tonight.

sloriot commented:
@freud14 I can make the change to `changes.html` if you add me to your collaborators for this repo.

freud14 commented:
@sloriot You're added. The branch is cone-spanners-ipelet.
